resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/Bulkhead.java

```diff
@@ -575,4 +601,54 @@ static Bulkhead of(String name, Supplier<BulkheadConfig> bulkheadConfigSupplier,
 
         EventPublisher onCallFinished(EventConsumer<BulkheadOnCallFinishedEvent> eventConsumer);
     }
+
+    /**
+     * This class decorates future with Bulkhead functionality around invocation.
+     *
+     * @param <T> of return type
+     */
+    final class BulkheadFuture<T> implements Future<T> {
+        final private Future<T> future;
+        final private OnceConsumer<Bulkhead> onceToBulkhead;
+
+        BulkheadFuture(Bulkhead bulkhead, Future<T> future) {
+            Objects.requireNonNull(future, "Non null Future is required to decorate");
+            this.onceToBulkhead = OnceConsumer.of(bulkhead);
+            this.future = future;
+
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return future.cancel(mayInterruptIfRunning);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return future.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return future.isDone();
+        }
+
+        @Override
+        public T get() throws InterruptedException, ExecutionException {
+            try {
+                return future.get();
+            }  finally {
+                onceToBulkhead.applyOnce(bh -> bh.onComplete());
```

Comment:
`cancel` or `timeout` exception should take it as Bulkead complete operation as permission was successfully acquired or should we release permission()?

Reply:
It seems to me we asked the same question over `CircuitBreakerFuture`, and the answer we came to was to assume that `get()` or `get(long, TimeUnit)` would be called on the returned `Future` at some point. So what you have here should be fine. If it's not, then we should revisit that question for the circuit breaker case as well.

Reply:
The permission was acquired and the supplier which returns the Future was invoked. The permission must not be released, if the future is canceled afterwards.

Reply:
Good point. If nobody invokes `future.get` the sempahore is never released. I would assume this is a bug in the application of the user then. 
Yes cancel and timout exceptions should also release the permission.